boxtree/tree_build.py

```diff
@@ -46,6 +56,240 @@ class MaxLevelsExceeded(RuntimeError):
     pass
 
 
+# {{{ utils for tree of boxes
+
+def resized_array(arr, new_size):
+    """Return a resized copy of the array. The new_size is a scalar which is
+    applied to the last dimension.
+    """
+    old_size = arr.shape[-1]
+    prefix = (slice(None), ) * (arr.ndim - 1)
+    if old_size >= new_size:
+        return arr[prefix + (slice(new_size), )].copy()
+    else:
+        new_shape = list(arr.shape)
+        new_shape[-1] = new_size
+        new_arr = np.zeros(new_shape, arr.dtype)
+        new_arr[prefix + (slice(old_size), )] = arr
+        return new_arr
+
+
+def vec_of_signs(dim, i):
+    """The sign vector is obtained by converting i to a dim-bit binary.
+    """
+    # e.g. bin(10) = '0b1010'
+    binary_digits = [int(bd) for bd in bin(i)[2:]]
```

Comment:
Probably better to do this without string conversion. See https://docs.python.org/3/library/stdtypes.html?highlight=bit_count#int.bit_count

Reply:
`bit_count` gives a total count of ones instead of each digit in the binary representation (from which the signs are computed).

Reply:
OK, different idea: Pyopencl has a `bitlog2`.

https://github.com/inducer/pyopencl/blob/71357fadc2745ac62f28f3384008ba61b400aa69/src/bitlog.hpp

Reply:
:shrug: I tried.